xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,424 @@
+        (
+            Buffer {
+                content: buffer.content >> length,
+                length: buffer.length - length
+            },
+            buffer.content & mask
+        )
+    }
+}
+
+#[test]
+fn test_buffer_pop() {
+    let buffer = Buffer { content: u32:0xDEADBEEF, length: u32:32 };
+    let (buffer, data) = buffer_pop(buffer, u32:16);
+    assert_eq(data, u32:0xBEEF);
+    assert_eq(buffer, Buffer { content: u32:0xDEAD, length: u32:16 });
+    let (buffer, data) = buffer_pop(buffer, u32:16);
+    assert_eq(data, u32:0xDEAD);
+    assert_eq(buffer, Buffer { content: u32:0, length: u32:0 });
+}
+
+// Returns `length` amount of data from a `buffer`, a new buffer with
+// the data removed and a positive status, if the buffer contains enough data.
+// Otherwise, it returns unmodified buffer, zeroed data field and error.
+// Since the Buffer structure acts as a simple FIFO, it pops the data in
+// the same order as they were added to the buffer.
+// The results are stored in the BufferResult structure.
+pub fn buffer_pop_checked<CAPACITY: u32> (buffer: Buffer<CAPACITY>, length: u32) -> (BufferResult<CAPACITY>, bits[CAPACITY]) {
+    if buffer_has_at_least(buffer, length) == false {
+        (
+            BufferResult { status: BufferStatus::NO_ENOUGH_DATA, buffer },
+            bits[CAPACITY]:0
+        )
+    } else {
+        let mask = (bits[CAPACITY]:1 << length) - bits[CAPACITY]:1;
+        (
+            BufferResult {
+                status: BufferStatus::OK,
+                buffer: Buffer {
+                    content: buffer.content >> length,
+                    length: buffer.length - length
+                }
+            },
+            buffer.content & mask
+        )
+    }
+}
+
+#[test]
+fn test_buffer_pop_checked() {
+    let buffer = Buffer { content: u32:0xDEADBEEF, length: u32:32 };
+
+    let (result1, data1) = buffer_pop_checked(buffer, u32:16);
+    assert_eq(result1, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { content: u32:0xDEAD, length: u32:16 }
+    });
+    assert_eq(data1, u32:0xBEEF);
+
+    let (result2, data2) = buffer_pop_checked(result1.buffer, u32:16);
+    assert_eq(result2, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { content: u32:0, length: u32:0 }
+    });
+    assert_eq(data2, u32:0xDEAD);
+
+    let (result3, data3) = buffer_pop_checked(result2.buffer, u32:16);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_DATA,
+        buffer: result2.buffer
+    });
+    assert_eq(data3, u32:0);
+}
+
+// Behaves like `buffer_pop` except that the length of the popped data can be
+// set using a DSIZE function parameter. For calls that need better error
+// handling, check `buffer_fixed_pop_checked`.
+pub fn buffer_fixed_pop<CAPACITY: u32, DSIZE: u32> (buffer: Buffer<CAPACITY>) -> (Buffer<CAPACITY>, bits[DSIZE]) {
+    let (buffer, value) = buffer_pop(buffer, DSIZE);
```

Comment:
you should be able to use a bit slice here given that `DSIZE` is a constant.

Reply:
I'm not sure if I understand you correctly but it would still be required to check if buffer has at least `DSIZE` bits of valid data inside. This is done in `buffer_pop()` and current implementation is universal and can be used when `length` is either a constant or regular variable. I think it is not entirely necessary to add a second implementation specified just for constants with slightly different handling of underlying bit array.

Reply:
@proppy I did some experiments on the unchecked variant of `buffer_pop()` functions. I've modified the implementation of `buffer_fixed_pop` to use bit slices as you suggested. The implementation was moved for minimal testing purposes into a separate file.
I've also put up a benchmarking setup for both variants of the function: it performs IR benchmarks, synthesis (with benchmark) and physical design flow targeting unscaled asap7 rev28 rvt. Target clock period is set to 750ps which is ~1.3GHz

Here are the results of IR benchmarks for both functions:
* [buffer_minimal_pop_ir_benchmark.txt](https://github.com/google/xls/files/13662887/buffer_minimal_pop_ir_benchmark.txt)
* [buffer_minimal_fixed_pop_ir_benchmark.txt](https://github.com/google/xls/files/13662888/buffer_minimal_fixed_pop_ir_benchmark.txt)

Synthesis benchmark is exactly the same for both functions:
```
Chip area for module '\buffer_minimal': 159.242760
Longest topological path in buffer_minimal (length=11):
Flop count: 352 objects.
Liberty: asap7-sc7p5t_rev28_rvt-ccs_ss_SS.lib
```

After running place and route flow for tested functions and inspecting timing reports from global routing I've noticed slack violations:
* [buffer_minimal_pop_place_and_route__global_routing.log](https://github.com/google/xls/files/13662991/buffer_minimal_pop_place_and_route__global_routing.log)
* [buffer_minimal_fixed_pop_place_and_route__global_routing.log](https://github.com/google/xls/files/13662989/buffer_minimal_fixed_pop_place_and_route__global_routing.log)

Then I decided to modify bazel rules for `place_and_route` to print additional reports after `detailed_routing` step is finished. Timings reported in this step are met:
* [buffer_minimal_pop_place_and_route__detailed_routing.log](https://github.com/google/xls/files/13662990/buffer_minimal_pop_place_and_route__detailed_routing.log)
* [buffer_minimal_fixed_pop_place_and_route__detailed_routing.log](https://github.com/google/xls/files/13662988/buffer_minimal_fixed_pop_place_and_route__detailed_routing.log)

The shortest path for both variants of the function has the same slack reported after detailed routing, while the slack for longest path is greater (48.078 vs 24.838) for the original implementation (using non-constant length of data to pop and masking).

I've pushed test setup to branch: https://github.com/antmicro/xls/tree/pcza/buffer-benchmarks

Reply:
did you also check if the codegen was different for the original fixed pop version calling the non fixed version?

Reply:
The difference between original `buffer_fixed_pop()` and one that was tested above is that I inlined the implementation of `buffer_pop()` there. Just to be sure I tested the original variant and the results are basically the same: wires have different numbers in names and one assign statement is split into 3 assigns like so:
```
-  assign p1_buffer__1_comb = {p1_or_reduce_258_comb ? {32'h0000_0000, p1_buffer_content__1_comb[127:32]} : p1_buffer_content__1_comb, p1_or_reduce_258_comb ? {p1_add_256_comb, p1_buffer_length__3_comb[4:0]} : p1_buffer_length__3_comb};
-  assign p1_value__1_comb = p1_buffer_content__1_comb[31:0] & {32{p1_or_reduce_258_comb}};
-  assign p1_tuple_271_comb = {p1_buffer__1_comb, p1_value__1_comb};
+  assign p1_tuple_203_comb = {{p1_or_reduce_190_comb ? {32'h0000_0000, p1_buffer_content__2_comb[127:32]} : p1_buffer_content__2_comb, p1_or_reduce_190_comb ? {p1_add_188_comb, p1_buffer_length__1_comb[4:0]} : p1_buffer_length__1_comb}, p1_buffer_content__2_comb[31:0] & {32{p1_or_reduce_190_comb}}};
```

Reply:
good to know that the constant arg gets optimized!